Wordsmithing repositories (#28)
uptane/deployment-considerations: repositories.md

1. Gather the tokens from each secondary ECU's version report.
2. Send the list of tokens to the Time Server to fetch the current time. The Time Server responds, as described in the [Time Server section](#time-server), by providing a cryptographic attestation of the last known time.
3. If the Time Server's response meets the criteria below, update the Primary ECU's clock and retain the Time Server's response for distribution to Secondary ECUs. If it fails to meet this criteria, discard the response and continue the procedure without an updated time.  The criteria for checking the Time Server's response are:
  - The signature over the Time Server's response is valid.
  - All the tokens provided to the Time Server are included in the response.
  - The time in the Time Server's response is later than the last time verified in this manner.

#### ECU Version Report

> The ECU version report from each Secondary will contain a token to be sent to the Time Server. To prevent a replay, this token SHOULD be unique for each update cycle. This stipulation should be possible to achieve as we expect updates to be relatively infrequent (e.g., due to limited number of write cycles), and that there will be a large number of possible tokens.

Author: Better, but it needs a little additional tweaks. How about this?
```suggestion
The ECU version report from each Secondary will contain a token to be sent to the Time Server. To prevent a replay, this token SHOULD be unique for each update cycle. As we expect that these updates will be relatively infrequent (e.g., due to a limited number of write cycles), there will be a sufficient number of tokens to make this possible"
```

> The ECU version report from each Secondary will contain a token to be sent to the Time Server. To prevent a replay, this token SHOULD be unique for each update cycle. This stipulation should be possible to achieve as we expect updates to be relatively infrequent (e.g., due to limited number of write cycles), and that there will be a large number of possible tokens.

Reviewer: Yep, looks good, other than the stray quote mark instead of a period at the end!

> The ECU version report from each Secondary will contain a token to be sent to the Time Server. To prevent a replay, this token SHOULD be unique for each update cycle. This stipulation should be possible to achieve as we expect updates to be relatively infrequent (e.g., due to limited number of write cycles), and that there will be a large number of possible tokens.

Author: ```suggestion
The ECU version report from each Secondary will contain a token to be sent to the Time Server. To prevent a replay, this token SHOULD be unique for each update cycle. As we expect that these updates will be relatively infrequent (e.g., due to a limited number of write cycles), there will be a sufficient number of tokens to make this possible.

```

> The payload of the ECU version report sent to the Director may contain the token sent to the Time Server. This token is part of the version report sent from Secondaries to the Primary. If the token is removed, the signature will not match.

Author: I found these few sentences rather snarly. I changed the might to me, because to me its use suggested also that it might not contain this token. If the token is needed to match the signature, then why is this not a should? What actually is the key point that needs to be made here? Any suggestions on how we can change this?

> The payload of the ECU version report sent to the Director may contain the token sent to the Time Server. This token is part of the version report sent from Secondaries to the Primary. If the token is removed, the signature will not match.

Reviewer: Here's my understanding: the previous paragraph describes how the tokens from Secondaries get to the Primary. This bit describes how the Primary sends those tokens and its own to the Time Server. One option is to include them as part of the ECU version report sent to the Director. They could be sent in some other way, but it seems simplest to do it that way. Hence, I think "MAY" is warranted, but I think we should be talking about "tokens", not just one "token". The second sentence ("This token is part of the version report sent from Secondaries to the Primary.") seems redundant with the preceding paragraph. And then the third sentence can be revised to make it clear that an advantage of putting the token(s) in the ECU version report is that the signature then includes that and protects against removing it. That paragraph can probably be merged with the preceding one. I'd like to suggest a change to do all that, but since it covers two paragraphs, I can't, and I don't have rights to push to your branch. I can open a PR against your own PR if that would help.

> The payload of the ECU version report sent to the Director may contain the token sent to the Time Server. This token is part of the version report sent from Secondaries to the Primary. If the token is removed, the signature will not match.

Author: How about I commit this PR and then you clarify this point in a separate one? It seems simpler that way.

> The payload of the ECU version report sent to the Director may contain the token sent to the Time Server. This token is part of the version report sent from Secondaries to the Primary. If the token is removed, the signature will not match.

Reviewer: Yeah, that's fine.

> The payload of the ECU version report sent to the Director may contain the token sent to the Time Server. This token is part of the version report sent from Secondaries to the Primary. If the token is removed, the signature will not match.

Reviewer: See https://github.com/uptane/deployment-considerations/pull/32.

**Figure 2.** *A proposed configuration of roles on the Director repository.*

> Unlike the Image repository, the Director repository does not delegate images. Therefore, the Director repository SHOULD contain only the Root, Timestamp, Snapshot, and Targets roles, as illustrated in Figure 2. In the rest of this section, we will discuss how metadata for each of these roles is produced.

Reviewer: If I'm gonna be really pedantic, "metadata" is plural, so "are" was correct. But did we decide to follow standard spoken usage instead? I don't really care either way.

Author: Yes. I think we let that one go.

#### Private API to update images and the inventory database

An OEM SHOULD define a private API for the Director repository, so that it is able to: (1) upload images, and (2) update the inventory database. This API is private in the sense that only the OEM should be able to perform these actions. 

This API SHOULD require authentication, so that each user is allowed to access only certain information. The OEM is free to use any authentication method, as long as it is suitably strong, such as [client certificates](https://blogs.msdn.microsoft.com/kaushal/2012/02/17/client-certificates-vs-server-certificates/), or [two-factor authentication](https://en.wikipedia.org/wiki/Multi-factor_authentication), such as a username coupled with a password, or an API key encrypted over TLS,.

> In order to allow automated processes on the Director repository to perform their respective functions, without also allowing any attackers who might compromise the repository to tamper with the inventory database, it is strongly RECOMMENDED that these processes should have some boundaries. That is, the automated processes SHOULD be able to read any record in the database, and write new records, but not be able to update or delete existing records.

Reviewer: ```suggestion
In order to allow automated processes on the Director repository to perform their respective functions, without also allowing any attackers who might compromise the repository to tamper with the inventory database, it is strongly RECOMMENDED that these processes should have some boundaries. That is, the automated processes SHOULD be able to read any record in the database and write new records, but SHOULD NOT be able to update or delete existing records.
```

* There SHOULD be a delegated Targets role for every tier-1 supplier, so that the OEM can:
  * limit the impact of a key compromise
  * precisely control which Targets metadata vehicles need to download.
  
The metadata for each tier-1 supplier MAY be signed by the OEM (e.g., supplier A), or the supplier itself (e.g., suppliers B and C). In turn, a tier-1 supplier MAY delegate images to members of its organization, such as supplier C, who has delegated a subset of its images to one of its developers, or its tier-2 suppliers who MAY delegate further to tier-3 suppliers.

Every delegation SHOULD be prefixed with the unique name of a tier-1 supplier, so that the filenames of images do not conflict with each other. Other than this constraint, a tier-1 supplier is free to name its images however it likes. For example, it MAY use the convention “supplier-X-ECU-Y-version-Z.img” to denote an image produced by supplier X, for ECU model Y, and with a version number Z.

#### Public API to download files

> An OEM SHOULD define a public API for Primaries to use when downloading metadata and images to the Image repository. This API can be defined however the OEM wishes.

Reviewer: ```suggestion
An OEM SHOULD define a public API for Primaries to use when downloading metadata and images from the Image repository. This API can be defined however the OEM wishes.
```
